base_layer/mmr/src/merkle_proof.rs

```diff
@@ -0,0 +1,275 @@
+    /// leaf position, call [MerkleProof::for_leaf_node] instead.
+    ///
+    /// The proof for the MMR consists of two parts:
+    /// a) A list of sibling node hashes starting from the candidate node and walking up the tree to the local root
+    /// (i.e. the root of the binary tree that the candidate node lives in.
+    /// b) A list of MMR peaks, excluding the local node hash.
+    /// The final Merkle proof is constructed by hashing all the peaks together (this is slightly different to how
+    /// other MMR implementations work).
+    pub fn for_node<D, B>(mmr: &MerkleMountainRange<D, B>, pos: usize) -> Result<MerkleProof, MerkleProofError>
+    where
+        D: Digest,
+        B: ArrayLike<Value = Hash>,
+    {
+        // check this pos is actually a leaf in the MMR
+        if !is_leaf(pos) {
+            return Err(MerkleProofError::NonLeafNode);
+        }
+
+        // check we actually have a hash in the MMR at this pos
+        mmr.get_node_hash(pos).ok_or(MerkleProofError::HashNotFound(pos))?;
+        let mmr_size = mmr.len();
+        let family_branch = family_branch(pos, mmr_size);
+
+        // Construct a vector of sibling hashes from the candidate node's position to the local peak
+        let path = family_branch
+            .iter()
+            .map(|(_, sibling)| {
+                mmr.get_node_hash(*sibling)
+                    .map(|v| v.clone())
+                    .ok_or(MerkleProofError::HashNotFound(*sibling))
+            })
+            .collect::<Result<_, _>>()?;
+
+        let peak_pos = match family_branch.last() {
+            Some(&(parent, _)) => parent,
+            None => pos,
+        };
+
+        // Get the peaks of the merkle trees, which are bagged together to form the root
+        // For the proof, we must leave out the local root for the candidate node
+        let peaks = find_peaks(mmr_size);
+        let mut peak_hashes = Vec::with_capacity(peaks.len() - 1);
+        for peak_index in peaks {
+            if peak_index != peak_pos {
+                let hash = mmr
+                    .get_node_hash(peak_index)
+                    .ok_or(MerkleProofError::HashNotFound(peak_index))?
+                    .clone();
+                peak_hashes.push(hash);
+            }
+        }
+        Ok(MerkleProof {
+            mmr_size,
+            path,
+            peaks: peak_hashes,
+        })
+    }
+
+    pub fn verify_leaf<D: Digest>(
+        &self,
+        root: &HashSlice,
+        hash: &HashSlice,
+        leaf_pos: usize,
+    ) -> Result<(), MerkleProofError>
+    {
+        let pos = leaf_index(leaf_pos);
+        self.verify::<D>(root, hash, pos)
+    }
+
+    /// Verifies the Merkle proof against the provided root hash, element and position in the MMR.
+    pub fn verify<D: Digest>(&self, root: &HashSlice, hash: &HashSlice, pos: usize) -> Result<(), MerkleProofError> {
```

Comment:
I'm not entirely sure when I would use this and when I'd use `verify_leaf`. Maybe some example code in the doc string would help

Reply:
Node and leaf numbering are different. Intuitively you'll want to use `verify_leaf` most of the time. I'll update the docstring accordingly